sw/airborne/modules/ctrl/optical_flow_landing.c

```diff
@@ -628,3 +1155,379 @@ void guidance_v_module_run(bool in_flight)
+  }
+}
+
+void learn_from_file(void)
+{
+  int i;
+  float fit_error;
+
+  // first load the texton distributions:
+  printf("Loading distribution\n");
+  load_texton_distribution();
+
+  // then learn from it:
+  printf("Learning!\n");
+  if (!RECURSIVE_LEARNING) {
+    fit_linear_model_OF(gains, text_dists, n_textons, n_read_samples, weights, &fit_error);
+    // fit_linear_model_OF(sonar_OF, text_dists, n_textons, n_read_samples, weights, &fit_error);
+  } else {
+    recursive_least_squares_batch(gains, text_dists, n_textons, n_read_samples, weights, &fit_error);
+  }
+
+  printf("Saving!\n");
+  // save the weights to a file:
+  save_weights();
+
+  printf("Learned! Fit error = %f\n", fit_error);
+
+  // free learning distributions:
+  for (i = 0; i < n_read_samples; i++) {
+    free(text_dists[i]);
+  }
+}
+
+/**
+ * Recursively fit a linear model from samples to target values - batch mode, possibly for initialization.
+ * @param[in] targets The target values
+ * @param[in] samples The samples / feature vectors
+ * @param[in] D The dimensionality of the samples
+ * @param[in] count The number of samples
+ * @param[out] parameters* Parameters of the linear fit
+ * @param[out] fit_error* Total error of the fit // TODO: relevant for RLS?
+ */
+void recursive_least_squares_batch(float *targets, float **samples, uint8_t D, uint16_t count, float *params,
+                                   float *fit_error)
+{
+  // TODO: potentially randomizing the sequence of the samples, as not to get a bias towards the later samples
+  // local vars for iterating, random numbers:
+  int sam, d;
+  uint8_t D_1 = D + 1;
+  float augmented_sample[D_1];
+  // augmented sample is used if the bias is used:
+  augmented_sample[D] = 1.0f;
+
+  // Initialize the weights with 0.0f:
+  for (d = 0; d < D_1; d++) {
+    weights[d] = 0.0f;
+  }
+
+  // Reset the P matrix to an identity matrix:
+  int i, j;
+  for (i = 0; i < n_textons + 1; i++) {
+    for (j = 0; j < n_textons + 1; j++) {
+      if (i == j) {
+        P_RLS[i][j] = 1.0f;
+      } else {
+        P_RLS[i][j] = 0.0f;
+      }
+    }
+  }
+
+  // give the samples one by one to the recursive procedure:
+  for (sam = 0; sam < count; sam++) {
+    if (!of_landing_ctrl.use_bias) {
+      recursive_least_squares(targets[sam], samples[sam], D, params);
+    } else {
+      for (d = 0; d < D; d++) {
+        augmented_sample[d] = samples[sam][d];
+      }
+      recursive_least_squares(targets[sam], augmented_sample, D_1, params);
+    }
+  }
+
+  // give the samples one by one to the recursive procedure to determine the fit on the training set:
+  float prediction;
+  float sum_abs_err = 0;
+  for (sam = 0; sam < count; sam++) {
+    prediction = predict_gain(samples[sam]);
+    sum_abs_err += fabs(prediction - targets[sam]);
+  }
+  (*fit_error) = sum_abs_err / count;
+}
+
+static inline void float_mat_div_scalar(float **o, float **a, float scalar, int m, int n)
```

Comment:
If relevant for other people, maybe these part can be moved to the math library ?

Reply:
added them as inline functions to pprz_algebra_float.h